File: tests/machine/x/cs/group_by_multi_join.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program {
    static void Main() {
        List<Nation> nations = new List<Nation> { new Nation { id = 1, name = "A" }, new Nation { id = 2, name = "B" } };
        List<Supplier> suppliers = new List<Supplier> { new Supplier { id = 1, nation = 1 }, new Supplier { id = 2, nation = 2 } };
        List<Partsupp> partsupp = new List<Partsupp> { new Partsupp { part = 100, supplier = 1, cost = 10.000000, qty = 2 }, new Partsupp { part = 100, supplier = 2, cost = 20.000000, qty = 1 }, new Partsupp { part = 200, supplier = 1, cost = 5.000000, qty = 3 } };
        List<Filtered> filtered = new Func<List<Filtered>>(() => {
    var _res = new List<Filtered>();
    foreach (var ps in partsupp) {
        foreach (var s in suppliers) {
            if (!((s.id == ps.supplier))) continue;
            foreach (var n in nations) {
                if (!((n.id == s.nation))) continue;
                if (!((n["name"] == "A"))) continue;
                _res.Add(new Filtered { part = ps.part, value = (ps.cost * ps.qty) });
            }
        }
    }
    return _res;
})();
        List<Grouped> grouped = _group_by(filtered, x => x.part).Select(g => new Grouped { part = g.Key, total = _sum(g.Items.Select(r => r.value).ToArray()) }).ToList();
        Console.WriteLine(JsonSerializer.Serialize(grouped));
    }
    public class Nation {
        public int id;
        public string name;
    }
    
    
    public class Supplier {
        public int id;
        public int nation;
    }
    
    
    public class Partsupp {
        public int part;
        public int supplier;
        public double cost;
        public int qty;
    }
    
    
    public class Filtered {
        public int part;
        public double value;
    }
    
    
    
    public class Grouped {
        public int part;
        public int total;
    }
    
    
    
    static double _sum(dynamic v) {
        if (v == null) return 0.0;
        double _sum = 0;
        foreach (var it in v) {
            _sum += Convert.ToDouble(it);
        }
        return _sum;
    }
    
    static List<_Group> _group_by(IEnumerable<dynamic> src, Func<dynamic, dynamic> keyfn) {
        var groups = new Dictionary<string, _Group>();
        var order = new List<string>();
        foreach (var it in src) {
            var key = keyfn(it);
            var ks = Convert.ToString(key);
            if (!groups.TryGetValue(ks, out var g)) {
                g = new _Group(key);
                groups[ks] = g;
                order.Add(ks);
            }
            g.Items.Add(it);
        }
        var res = new List<_Group>();
        foreach (var k in order) res.Add(groups[k]);
        return res;
    }
    
}

File: tests/machine/x/scala/group_by_multi_join.scala
object group_by_multi_join {
  case class Filtered(part: Int, value: Double)
  case class Grouped(part: Int, total: Int)
  case class Nation(id: Int, name: String)
  case class Partsupp(part: Int, supplier: Int, cost: Double, qty: Int)
  case class Supplier(id: Int, nation: Int)

  case class _Group[K,T](key: K, items: List[T]) extends Iterable[T] { def iterator: Iterator[T] = items.iterator }

  val nations = List[Nation](Nation(id = 1, name = "A"), Nation(id = 2, name = "B"))
  val suppliers = List[Supplier](Supplier(id = 1, nation = 1), Supplier(id = 2, nation = 2))
  val partsupp = List[Partsupp](Partsupp(part = 100, supplier = 1, cost = 10, qty = 2), Partsupp(part = 100, supplier = 2, cost = 20, qty = 1), Partsupp(part = 200, supplier = 1, cost = 5, qty = 3))
  val filtered = for { ps <- partsupp; s <- suppliers; if (s.id).asInstanceOf[Int] == ps.supplier; n <- nations; if (n.id).asInstanceOf[Int] == s.nation; if n.name == "A" } yield Filtered(part = ps.part, value = ps.cost * ps.qty)
  val grouped = ((for { x <- filtered } yield (x.part, x)).groupBy(_._1).map{ case(k,list) => _Group(k, list.map(_._2)) }.toList).map{ g => Grouped(part = g.key, total = (for { r <- g } yield r.value).sum) }.toList
  def main(args: Array[String]): Unit = {
    println((grouped))
  }
}
